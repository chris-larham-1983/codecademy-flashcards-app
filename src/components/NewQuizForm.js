import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import { v4 as uuidv4 } from "uuid";
import ROUTES from "../app/routes";
import { useSelector, useDispatch } from 'react-redux';
import { topicsSelector } from '../features/topics/topicsSlice';
import { addQuizThunk } from '../features/quizzes/quizzesSlice';
import { addCard } from '../features/cards/cardsSlice';

export default function NewQuizForm() {
  const [name, setName] = useState("");
  const [cards, setCards] = useState([]);
  const [topicId, setTopicId] = useState("");
  const history = useHistory();
  const topics = useSelector(topicsSelector);
  const dispatch = useDispatch();

  const handleSubmit = (e) => {
    e.preventDefault();
    if (name.trim().length === 0) {
      document.getElementById('quizNameError').removeAttribute("hidden");
      return;
    }

    if(topicId === '') {
      document.getElementById('topicIdError').removeAttribute("hidden");
      return;
    }

    const cardIds = [];

    cards.forEach((card) => {
      let cardId = uuidv4();
      cardIds.push(cardId);
      dispatch(addCard({
        ...card,
        id: cardId
      }));
    });

    dispatch(addQuizThunk({
          id: uuidv4(),
          name: name,
          topicId: topicId,
          cardIds: cardIds
        })
    );

    history.push(ROUTES.quizzesRoute());
  };

  const addCardInputs = (e) => {
    e.preventDefault();
    setCards(cards.concat({ front: "", back: "" }));
  };

  const removeCard = (e, index) => {
    e.preventDefault();
    setCards(cards.filter((card, i) => index !== i));
  };

  const updateCardState = (index, side, value) => {
    const newCards = cards.slice();
    newCards[index][side] = value;
    setCards(newCards);
  };

  return (
      <section>
        <h1>Create a new quiz</h1>
        <form onSubmit={handleSubmit}>
          <input
              id="quiz-name"
              value={name}
              onChange={(e) => {
                  setName(e.currentTarget.value);
                  if(e.currentTarget.value.trim() === '') {
                      document.getElementById("quizNameError").removeAttribute("hidden");
                  } else {
                      document.getElementById("quizNameError").setAttribute("hidden", "true");
                  }}}
              placeholder="Quiz Title"
          />
          <p id="quizNameError" hidden={true}>Please enter a quiz name.</p>
          <select
              id="quiz-topic"
              onChange={(e) => {
                setTopicId(e.currentTarget.value);
                if(e.currentTarget.value.trim() === '') {
                  document.getElementById("topicIdError").removeAttribute("hidden");
                } else {
                  document.getElementById("topicIdError").setAttribute("hidden", "true");
                }}}
              placeholder="Topic"
          >
            <option value="">Topic</option>
            {Object.values(topics).map((topic) => (
                <option key={topic.id} value={topic.id}>
                  {topic.name}
                </option>
            ))}
          </select>
          <p id='topicIdError' hidden={true}>Please select a valid topic.  To create a topic, click the "Topics" tab.</p>
          {cards.map((card, index) => (
              <div key={index} className="card-front-back">
                <input
                    id={`card-front-${index}`}
                    value={cards[index].front}
                    onChange={(e) =>
                        updateCardState(index, "front", e.currentTarget.value)
                    }
                    placeholder="Front"
                />

                <input
                    id={`card-back-${index}`}
                    value={cards[index].back}
                    onChange={(e) =>
                        updateCardState(index, "back", e.currentTarget.value)
                    }
                    placeholder="Back"
                />

                <button
                    onClick={(e) => removeCard(e, index)}
                    className="remove-card-button"
                >
                  Remove Card
                </button>
              </div>
          ))}
          <div className="actions-container">
            <button className="addCard" onClick={addCardInputs}>Add a Card</button>
            <button>Create Quiz</button>
          </div>
        </form>
      </section>
  );
}
