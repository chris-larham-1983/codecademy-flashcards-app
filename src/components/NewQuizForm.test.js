import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import NewQuizForm from './NewQuizForm';
import { render, fireEvent } from '@testing-library/react';
import { screen } from '@testing-library/dom';

const mockStore = configureStore([]);

describe('<NewQuizForm /> component', () => {
    let store;
    let component;

    beforeEach(() => {
        store = mockStore({
            topics: {
                topics: {
                    '1': {
                        id: '1',
                        name: 'Topic 1',
                        icon: 'Icon URL 1',
                        quizIds: ['2'],
                    }
                }
            },
            quizzes: {
                quizzes: {
                    '2': {
                        id: '2',
                        topicId: '1',
                        name: 'Quiz 1 for Topic 1',
                        cardIds: ['1', '2', '3'],
                    }
                }
            }
        });

        component = renderer.create(
            <Provider store={store}>
                <NewQuizForm />
            </Provider>
        );
    });

    describe('rendering', () => {
        it('should render the component without crashing', () => {
            expect(component.toJSON()).toBeTruthy();
        });

        it('should render with given state from the Redux store', () => {
            expect(component.toJSON()).toMatchSnapshot();
        });

        it('should render an h1 heading element', () => {
            expect(component.root.findByType('h1')).toBeTruthy();
            expect(component.root.findByType('h1').children).toStrictEqual(['Create a new quiz']);
        });

        it('should render a form', () => {
            expect(component.root.findByType('form')).toBeTruthy();
        });

        it('should render an "Add a Card" button', () => {
            expect(component.root.findAllByType('button')[0].children).toStrictEqual(['Add a Card']);
        });

        it('should render a "Create Quiz" button', () => {
            expect(component.root.findAllByType('button')[1].children).toStrictEqual(['Create Quiz']);
        });

        it('should not initially display a paragraph that informs the User of the need to enter a quiz name', () => {
            expect(component.root.findByProps({ id: "quizNameError" })).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[0]).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[0].children).toStrictEqual(['Please enter a quiz name.']);
        });

        it('should not initially display a paragraph that informs the User of the need to select a topic', () => {
            expect(component.root.findByProps({ id: "topicIdError"})).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[1]).toBeTruthy();
            expect(component.root.findAllByProps({ hidden: true })[1].children).toStrictEqual(['Please select a valid topic.  To create a topic, click the "Topics" tab.']);
        });

        describe('when the User enters a quiz name', () => {
            let store;
            let component;

            beforeEach(() => {
                store = mockStore({
                    topics: {
                        topics: {
                            '1': {
                                id: '1',
                                name: 'Topic 1',
                                icon: 'Icon URL 1',
                                quizIds: ['2'],
                            }
                        }
                    },
                    quizzes: {
                        quizzes: {
                            '2': {
                                id: '2',
                                topicId: '1',
                                name: 'Quiz 1 for Topic 1',
                                cardIds: ['1', '2', '3'],
                            }
                        }
                    }
                });

                component = render(
                    <Provider store={store}>
                        <NewQuizForm />
                    </Provider>
                );
            });


            it('should hide the paragraph that instructs the User to enter a quiz name', () => {
                const createQuizButton = screen.getByText(/Create Quiz/i);
                const quizNameInput = screen.getByPlaceholderText(/Quiz Title/i);
                expect(screen.getByText(/Please enter a quiz name./i)).not.toBeVisible(); //descriptive paragraph should initially be hidden
                fireEvent.click(createQuizButton); //click the button with no quiz name entered
                expect(screen.getByText(/Please enter a quiz name./i)).toBeVisible(); //expect the descriptive paragraph to be visible
                fireEvent.change(quizNameInput, { target: { value: 'New Quiz Name' }}); //change the value of the quiz name input element to 'New Quiz Name'
                expect(screen.getByText(/Please enter a quiz name./i)).not.toBeVisible(); //expect the descriptive paragraph to no longer be visible
            });
        });

        describe('when the User selects a valid topic', () => {
            let store;
            let component;

            beforeEach(() => {
                store = mockStore({
                    topics: {
                        topics: {
                            '1': {
                                id: '1',
                                name: 'Topic 1',
                                icon: 'Icon URL 1',
                                quizIds: ['2'],
                            }
                        }
                    },
                    quizzes: {
                        quizzes: {
                            '2': {
                                id: '2',
                                topicId: '1',
                                name: 'Quiz 1 for Topic 1',
                                cardIds: ['1', '2', '3'],
                            }
                        }
                    }
                });

                component = render(
                    <Provider store={store}>
                        <NewQuizForm />
                    </Provider>
                );
            });
            it('should hide the descriptive paragraph that instructs the User to select a valid topic', () => {
                const createQuizButton = screen.getByText(/Create Quiz/i);
                const quizNameInput = screen.getByPlaceholderText(/Quiz Title/i);
                const topicSelector = screen.getByPlaceholderText(/Topic/);
                expect(topicSelector).toBeInTheDocument();
                fireEvent.change(quizNameInput, { target: { value: 'New Quiz Name' }}); //change the value of the quiz name input element to 'New Quiz Name'
                fireEvent.click(createQuizButton); //click the button with a quiz name entered BUT no quiz topic selected
                expect(screen.getByText(/Please enter a quiz name./i)).not.toBeVisible(); //expect the descriptive paragraph related to the quiz name not to be visible
                expect(screen.getByText(/Please select a valid topic. To create a topic, click the "Topics" tab./i)).toBeVisible(); //expect the descriptive paragraph related to the topic to be visible
                topicSelector[1].selected = true; //set the topicSelector value to 'Topic 1'
                fireEvent.change(topicSelector); //fire a change event on the topicSelector
                expect(screen.getByText(/Please select a valid topic. To create a topic, click the "Topics" tab./i)).not.toBeVisible(); //expect the descriptive paragraph related to the topic icon NOT to be visible
            });
        });
    });
});