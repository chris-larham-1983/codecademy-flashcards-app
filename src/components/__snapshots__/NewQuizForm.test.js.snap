// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewQuizForm /> component rendering should render with given state from the Redux store 1`] = `
<section>
  <h1>
    Create a new quiz
  </h1>
  <form
    onSubmit={[Function]}
  >
    <input
      id="quiz-name"
      onChange={[Function]}
      placeholder="Quiz Title"
      value=""
    />
    <p
      hidden={true}
      id="quizNameError"
    >
      Please enter a quiz name.
    </p>
    <select
      id="quiz-topic"
      onChange={[Function]}
      placeholder="Topic"
    >
      <option
        value=""
      >
        Topic
      </option>
      <option
        value="1"
      >
        Topic 1
      </option>
    </select>
    <p
      hidden={true}
      id="topicIdError"
    >
      Please select a valid topic.  To create a topic, click the "Topics" tab.
    </p>
    <div
      className="actions-container"
    >
      <button
        className="addCard"
        onClick={[Function]}
      >
        Add a Card
      </button>
      <button>
        Create Quiz
      </button>
    </div>
  </form>
</section>
`;
